src/blob/index/bptree/core.rs

```diff
@@ -198,31 +198,37 @@ impl<K> BPTreeFileIndex<K>
 where
     for<'a> K: Key<'a> + 'static,
 {
-    async fn find_leaf_node(&self, key: &K, mut offset: u64, buf: &mut [u8]) -> Result<u64> {
+    async fn find_leaf_node(
+        &self,
+        key: &K,
+        mut offset: u64,
+        mut buf: BytesMut,
+    ) -> Result<(BytesMut, u64)> {
         while offset < self.metadata.leaves_offset {
             offset = if offset == self.metadata.tree_offset {
                 Node::key_offset_serialized(&self.root_node, key)?
             } else {
-                self.file.read_at(buf, offset).await?;
-                Node::key_offset_serialized(buf, key)?
+                buf = self.file.read_at(buf, offset).await?;
+                Node::key_offset_serialized(&buf, key)?
             };
         }
-        Ok(offset)
+        Ok((buf, offset))
     }
 
     async fn read_header(
         &self,
         leaf_offset: u64,
         key: &K,
-        buf: &mut [u8],
+        mut buf: BytesMut,
     ) -> Result<Option<RecordHeader>> {
         let buf_size = self.leaf_node_buf_size(leaf_offset);
-        self.file.read_at(&mut buf[..buf_size], leaf_offset).await?;
+        buf.resize(buf_size, 0);
```

Comment:
There was expectation before, that `buf_size <= BLOCK_SIZE`. I suggest to add an explicit check for that and return error if `buf_size > BLOCK_SIZE`